File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
class Solution {
    private int count = 0;
    private int ans = -1;

    public int kthSmallest(TreeNode root, int k) {
        inorder(root, k);
        return ans;
    }

    private void inorder(TreeNode root, int k) {
        if (root == null || count >= k) return;

        inorder(root.left, k);

        count++;
        if (count == k) {
            ans = root.val;
            return;
        }

        inorder(root.right, k);
    }
}

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
class Solution {
public:
    void inorder(TreeNode* root, int& k, int& ans) {
        if (!root || k <= 0) return;

        inorder(root->left, k, ans);

        if (k > 0) {
            k--;

            if (k == 0) {
                ans = root->val;
                return;
            }

            inorder(root->right, k, ans);
        }
    }

    int kthSmallest(TreeNode* root, int k) {
        int ans = -1;
        inorder(root, k, ans);
        
        return ans;
    }
};